boring-sys/patches/boring-pq.patch

```diff
@@ -154,3178 +172,3184 @@ index 4c0048e1d..f6ea5c40f 100644
++        r[j + len] = fqmul(zeta, r[j + len]);
++      }
++    }
+   }
++
++  for(j = 0; j < 256; j++)
++    r[j] = fqmul(r[j], f);
+ }
  
--      int chunk_bits = bits - element_bits_done;
--      if (chunk_bits > in_byte_bits_left) {
--        chunk_bits = in_byte_bits_left;
--      }
+-static void scalar_add(scalar *lhs, const scalar *rhs) {
+-  for (int i = 0; i < DEGREE; i++) {
+-    lhs->c[i] = reduce_once(lhs->c[i] + rhs->c[i]);
 +/*************************************************
-+* Name:        unpack_pk
++* Name:        basemul
 +*
-+* Description: De-serialize public key from a byte array;
-+*              approximate inverse of pack_pk
++* Description: Multiplication of polynomials in Zq[X]/(X^2-zeta)
++*              used for multiplication of elements in Rq in NTT domain
 +*
-+* Arguments:   - polyvec *pk: pointer to output public-key polynomial vector
-+*              - uint8_t *seed: pointer to output seed to generate matrix A
-+*              - const uint8_t *packedpk: pointer to input serialized public key
++* Arguments:   - int16_t r[2]: pointer to the output polynomial
++*              - const int16_t a[2]: pointer to the first factor
++*              - const int16_t b[2]: pointer to the second factor
++*              - int16_t zeta: integer defining the reduction polynomial
 +**************************************************/
-+static void unpack_pk(polyvec *pk,
-+                      uint8_t seed[KYBER_SYMBYTES],
-+                      const uint8_t packedpk[KYBER_INDCPA_PUBLICKEYBYTES])
++static void basemul(int16_t r[2], const int16_t a[2], const int16_t b[2], int16_t zeta)
 +{
-+  size_t i;
-+  polyvec_frombytes(pk, packedpk);
-+  for(i=0;i<KYBER_SYMBYTES;i++)
-+    seed[i] = packedpk[i+KYBER_POLYVECBYTES];
++  r[0]  = fqmul(a[1], b[1]);
++  r[0]  = fqmul(r[0], zeta);
++  r[0] += fqmul(a[0], b[0]);
++  r[1]  = fqmul(a[0], b[1]);
++  r[1] += fqmul(a[1], b[0]);
 +}
- 
--      element |= (in_byte & kMasks[chunk_bits - 1]) << element_bits_done;
--      in_byte_bits_left -= chunk_bits;
--      in_byte >>= chunk_bits;
++
++//
++// poly.c
++//
++
 +/*************************************************
-+* Name:        pack_sk
++* Name:        poly_compress
 +*
-+* Description: Serialize the secret key
++* Description: Compression and subsequent serialization of a polynomial
 +*
-+* Arguments:   - uint8_t *r: pointer to output serialized secret key
-+*              - polyvec *sk: pointer to input vector of polynomials (secret key)
++* Arguments:   - uint8_t *r: pointer to output byte array
++*                            (of length KYBER_POLYCOMPRESSEDBYTES)
++*              - const poly *a: pointer to input polynomial
 +**************************************************/
-+static void pack_sk(uint8_t r[KYBER_INDCPA_SECRETKEYBYTES], polyvec *sk)
++static void poly_compress(uint8_t r[KYBER_POLYCOMPRESSEDBYTES], const poly *a)
 +{
-+  polyvec_tobytes(r, sk);
-+}
++  unsigned int i,j;
++  int16_t u;
++  uint32_t d0;
```

Comment:
You use `uint64_t` for the same variable below in `polyvec_compress()`. Is `uint32_t` large enough here?

Reply:
From the name (taken from upstream) you would think `polyvec_compress` is just a component-wise version of `poly_vec`, which it's not. For Kyber768 the former leaves 10 bits per coefficient (the `d_u` parameter), whereas the latter leaves 4 (the `d_v` parameter).  Because the numbers are bigger, we also need a more accurate Barrett approximant for the division by Q. The intermediate values in the latter can grow as large as 2^42, whereas they stay below 32 in the former.